File: fixilya-backend/fixilya-microservices/services/call-service/src/services/call.service.js
'use strict';

const path = require('path');
const admin = require(path.join(__dirname, '../../../../shared/config/firebase'));
const { AppError } = require(path.join(__dirname, '../../../../shared/utils/appError'));
const logger = require(path.join(__dirname, '../../../../shared/utils/logger'));
const { buildRtcToken, ROLE_PUBLISHER } = require('../utils/agoraToken');

const db = admin.firestore();
const CALLS_COLLECTION = 'calls';
const BOOKINGS_COLLECTION = 'bookings';

// Token TTL: 1 hour, clamped between 60 s and 24 h
const TOKEN_TTL = 3600;

class CallService {
  // ─────────────────────────────────────────────
  // INITIATE
  // ─────────────────────────────────────────────

  /**
   * Creates a new call session.
   *
   * Security:
   * - Verifies callerId and calleeId share a confirmed/in_progress booking.
   * - Generates caller Agora token server-side (never client-generated).
   * - Server writes the Firestore call doc (client cannot forge calls).
   *
   * @param {string} callerId  Firebase UID of the caller
   * @param {object} opts
   * @param {string} opts.calleeId
   * @param {string} opts.callerName
   * @param {string} opts.calleeName
   * @param {string} [opts.callerPicture]
   * @param {string} [opts.calleePicture]
   * @returns {{ callId: string, callerToken: string }}
   */
  async initiateCall(callerId, { calleeId, callerName, calleeName, callerPicture, calleePicture }) {
    const appId = process.env.AGORA_APP_ID;
    const appCert = process.env.AGORA_APP_CERTIFICATE;

    if (!appId || !appCert) {
      throw new AppError('Agora is not configured on this server', 503);
    }

    if (!calleeId || calleeId === callerId) {
      throw new AppError('Invalid callee', 400);
    }

    // ── Booking authorization ──────────────────────────────────────────────
    // Check both directions: client→handyman and handyman→client
    const [q1, q2] = await Promise.all([
      db.collection(BOOKINGS_COLLECTION)
        .where('clientId',   '==', callerId)
        .where('handymanId', '==', calleeId)
        .where('status', 'in', ['confirmed', 'in_progress'])
        .limit(1)
        .get(),
      db.collection(BOOKINGS_COLLECTION)
        .where('handymanId', '==', callerId)
        .where('clientId',   '==', calleeId)
        .where('status', 'in', ['confirmed', 'in_progress'])
        .limit(1)
        .get(),
    ]);

    if (q1.empty && q2.empty) {
      logger.warn(`[CallService.initiateCall] No active booking between ${callerId} and ${calleeId}`);
      throw new AppError('No active booking found between caller and callee', 403);
    }

    // ── Generate caller token ──────────────────────────────────────────────
    const callDoc = db.collection(CALLS_COLLECTION).doc();
    const callId = callDoc.id;

    const callerToken = buildRtcToken(appId, appCert, callId, 0, ROLE_PUBLISHER, TOKEN_TTL);

    // ── Write call document (server-side only) ────────────────────────────
    await callDoc.set({
      callId,
      callerId,
      calleeId,
      callerName:    callerName   || 'Caller',
      calleeName:    calleeName   || 'Callee',
      callerPicture: callerPicture || null,
      calleePicture: calleePicture || null,
      channelName:   callId,  // Agora channel name == call document ID
      status:        'ringing',
      createdAt:     admin.firestore.FieldValue.serverTimestamp(),
      updatedAt:     admin.firestore.FieldValue.serverTimestamp(),
    });

    logger.info(`[CallService.initiateCall] callId=${callId} caller=${callerId} callee=${calleeId}`);

    return { callId, callerToken };
  }

  // ─────────────────────────────────────────────
  // ACCEPT
  // ─────────────────────────────────────────────

  /**
   * Validates the callee, transitions status to 'active', and returns
   * a fresh Agora token for the callee to join the channel.
   *
   * Security:
   * - Verifies req.uid === call.calleeId (enforced by controller).
   * - Returns 404 if call not found, 409 if already active/ended.
   * - Generates a brand-new token — nothing sensitive is stored in Firestore.
   *
   * @param {string} calleeUid  Firebase UID from verified JWT
   * @param {string} callId     Firestore document ID
   * @returns {{ calleeToken: string }}
   */
  async acceptCall(calleeUid, callId) {
    const appId = process.env.AGORA_APP_ID;
    const appCert = process.env.AGORA_APP_CERTIFICATE;

    if (!appId || !appCert) {
      throw new AppError('Agora is not configured on this server', 503);
    }

    const callSnap = await db.collection(CALLS_COLLECTION).doc(callId).get();
    if (!callSnap.exists) {
      throw new AppError('Call not found', 404);
    }

    const call = callSnap.data();

    if (call.calleeId !== calleeUid) {
      throw new AppError('Forbidden', 403);
    }

    if (call.status !== 'ringing') {
      throw new AppError(`Call is already ${call.status}`, 409);
    }

    // Generate a fresh callee token — never stored in Firestore
    const calleeToken = buildRtcToken(appId, appCert, callId, 0, ROLE_PUBLISHER, TOKEN_TTL);

    await db.collection(CALLS_COLLECTION).doc(callId).update({
      status:    'active',
      updatedAt: admin.firestore.FieldValue.serverTimestamp(),
    });

    logger.info(`[CallService.acceptCall] callId=${callId} callee=${calleeUid} → active`);

    return { calleeToken };
  }
}

module.exports = new CallService();

File: lib/services/call_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:dio/dio.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:fixilya_app/core/config/app_config.dart';
import 'package:fixilya_app/services/api_client.dart';
import 'package:flutter/foundation.dart';

void debugLog(String msg) {
  if (kDebugMode) debugPrint('[CallService] $msg');
}

/// Manages call signaling through Firestore + Agora RTC.
///
/// Flow:
///   1. Client presses "Call" → [initiateCall] calls call-service which writes the Firestore doc
///   2. Handyman app listens to /calls with status == 'ringing' for their uid
///   3. Handyman accepts → [acceptCall] fetches callee token from call-service (sets status = 'active')
///   4. Handyman rejects → [rejectCall] sets status = 'rejected'
///   5. Either side hangs up → [endCall] sets status = 'ended'
///
/// The Agora channel name IS the callId, so both sides join the same channel.
/// Tokens are generated server-side — never client-side.
class CallService {
  static final CallService _instance = CallService._internal();
  factory CallService() => _instance;
  CallService._internal();

  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  final FirebaseAuth _auth = FirebaseAuth.instance;

  // ─────────────────────────────────────────────
  // AGORA CONFIG
  // ─────────────────────────────────────────────

  /// App ID read from AppConfig (populated via --dart-define in production).
  static String get agoraAppId => AppConfig.agoraAppId;

  // ─────────────────────────────────────────────

  String get _currentUid => _auth.currentUser!.uid;

  /// Creates a call session via call-service.
  ///
  /// The server validates that caller and callee share an active booking,
  /// writes the Firestore call doc, and returns the caller's Agora token.
  ///
  /// Returns `{'callId': String, 'token': String?}`.
  Future<Map<String, String?>> initiateCall({
    required String calleeId,
    required String calleeName,
    required String callerName,
    String? calleePicture,
    String? callerPicture,
  }) async {
    try {
      final response = await ApiClient().callDio.post(
        '/calls/initiate',
        data: {
          'calleeId':      calleeId,
          'calleeName':    calleeName,
          'callerName':    callerName,
          if (calleePicture != null) 'calleePicture': calleePicture,
          if (callerPicture != null) 'callerPicture': callerPicture,
        },
      );
      if (response.statusCode == 201 && response.data['success'] == true) {
        final data = response.data['data'] as Map<String, dynamic>;
        return {
          'callId': data['callId'] as String,
          'token':  data['callerToken'] as String?,
        };
      }
      throw Exception('Unexpected response from call-service: ${response.statusCode}');
    } on DioException catch (e) {
      debugLog('initiateCall error: ${e.type} — ${e.response?.data}');
      rethrow;
    }
  }

  /// Fetches the callee's Agora token from call-service (also sets status → active).
  ///
  /// Returns the Agora token string, or null if the request fails.
  Future<String?> acceptCall(String callId) async {
    try {
      final response = await ApiClient().callDio.get('/calls/$callId/accept');
      if (response.statusCode == 200 && response.data['success'] == true) {
        return response.data['data']['calleeToken'] as String?;
      }
      return null;
    } on DioException catch (e) {
      debugLog('acceptCall error: ${e.type}');
      return null;
    }
  }

  /// Handyman rejects the call.
  Future<void> rejectCall(String callId) async {
    await _firestore.collection('calls').doc(callId).update({
      'status': 'rejected',
      'updatedAt': FieldValue.serverTimestamp(),
    });
  }

  /// Either side ends the call.
  Future<void> endCall(String callId) async {
    await _firestore.collection('calls').doc(callId).update({
      'status': 'ended',
      'updatedAt': FieldValue.serverTimestamp(),
    });
  }

  /// Mark call as missed (called when ringing timeout expires with no answer).
  Future<void> markMissed(String callId) async {
    await _firestore.collection('calls').doc(callId).update({
      'status': 'missed',
      'updatedAt': FieldValue.serverTimestamp(),
    });
  }

  /// Stream the call document for real-time status updates.
  Stream<DocumentSnapshot<Map<String, dynamic>>> watchCall(String callId) {
    return _firestore.collection('calls').doc(callId).snapshots();
  }

  /// Stream for incoming calls targeting this user (handyman listening).
  Stream<QuerySnapshot<Map<String, dynamic>>> watchIncomingCalls() {
    return _firestore
        .collection('calls')
        .where('calleeId', isEqualTo: _currentUid)
        .where('status', isEqualTo: 'ringing')
        .orderBy('createdAt', descending: true)
        .limit(1)
        .snapshots();
  }

  /// Fetch a single call document.
  Future<Map<String, dynamic>?> getCall(String callId) async {
    final doc = await _firestore.collection('calls').doc(callId).get();
    return doc.data();
  }
}
